

![[Node.js 동기&비동기1.png]]


위 그림의 왼쪽처럼 **하나의 작업이 끝날 때까지 기다렸다가 다음 작업이 처리하는 방식**을 **동기 처리 방식(synchronous)**라고 하고, 그림의 오른쪽처럼 **병렬로 동시에 여러 가지 일을 처리하는 방식**을 **비동기 처리 방식(asynchronous)**라고 한다.  
즉, 비동기 처리 방식은 여러 작업을 동시에 처리함으로써 작업 처리를 좀 더 효율적으로 할 수 있게 만들어 준다.

Node.js는 비동기 처리를 하기 위한 아주 좋은 기능을 제공한다. 그런데 애플리케이션이 비동기 처리 방식으로 동작하면 **효율적이지만 매우 복잡**하다.


Node.js가 비동기처리를 하는 함수나 메소드(더 있을수 있음)

> 1. setTimeout
> 2. setInterval
> 3. Ajax
> 4. Request
> 5. fetch
> 6. axios
> 7. 데이터베이스 쿼리문


# 동기, 비동기 예제

Node.js에서 보통 'Sync'라는 단어가 있으면 동기 처리 방식으로 작동하는 기능이고, 'Sync'가 없으면 비동기 처리 방식으로 동작하는 기능이다. 즉, 기본 이름에 **'Sync'**가 붙으면 **해당 기능을 동기 처리 방식으로 처리한다는 의미**이다.

### 콜백함수

콜백함수란 나중에 실행되는 함수로 비동기 작업이 끝나고 난뒤 실행되는 함수이다.

```js
setTimeout((arg)=>{
    console.log(arg)
    
    setTimeout((arg)=>{
        console.log(arg)

        setTimeout((arg)=>{
            console.log(arg)

            setTimeout((arg)=>{
                console.log(arg)

            }, 1000, "네번째 콜백")
        }, 1000, "세번째 콜백")
    }, 1000, "두번째 콜백")
}, 1000, "첫번째 콜백")
```

이러한 콜백지옥의 문제로 가독성도 떨어져 ES6에 promise 객체가 나왔다.

---

### promise(resolve, reject) + then, catch

프로미스 객체는 콜백함수의 콜백지옥을 해결하기 위해 나왔다.

```js
const test = new Promise((resolve, reject)=>{
    if(resolve){
        resolve("성공")
    }
    if(reject){
        reject("실패")
    }
});
  
console.log(test) // Promise { <pending> }
```

하지만 프로미스를 써서 비동기처리를 하게되면 반환값이 프로미스 객체로 반환되게 되는데 이때 사용되는것이 then과 catch이다.

```js
const test = new Promise((resolve, reject)=>{
    if(resolve){
        resolve("성공")
    }
    if(reject){
        reject("실패")
    }
});
  
test.then(result=>{
    return console.log(result)
	})
  .catch(err=>{
 	return console.log(result)
	})
	// "성공"
```

프로미스를 사용해서 성공시에는 then에 인자로 닮겨서 사용이 가능하고 실패시 catch에 담겨서 예외처리를 할수있다.

**프로미스객체의 상태 3가지**

> Pending(대기) : 미완료 상태  
> Fulfilled(이행) : 완료 상태  
> Rejected(실패) : 실패 상태

```js
function promiseTest() { 
    return new Promise((resolve, reject) => {
        try{
            // 함수 작성
            setTimeout(() => { resolve('Success') },1000)           
        } catch (err) {
            reject('Fail') // 또는 throw new Error('error')
        }
    })
}

function main() {
	promiseTest().then(data => {
        console.log(data)
    }).catch(err => {
        console.log('ERROR : ' + err)
    })
    console.log('End')
}

main()
```

---

### async & await

ES2017에 나온 문법이며 프로미스 객체를 좀더 사용하기 쉽게 만들어진게 async & await이다.
사용법은 function 앞에 async 붙여주고 그 함수 안에서 실행하려는 함수 앞에 await를 붙여 사용한다. 시스템에서 코드를 실행하다가 await 함수 실행을 만나면 바로 실행된다. await는 async가 붙은 함수 안에서만 작동한다.


```js
const test = async () => {
    const test =  await 비동기함수
}
```


```js
function promiseTest() { 
    return new Promise((resolve,reject) => {
        setTimeout(() => {
            console.log('Success')
            resolve()
        },1000)
    })
}


async function main() {
    try {
        await promiseTest()
        console.log('End')
    } catch(err) {
        console.log('ERROR : ' + err.message)
    }   
}

main()
```


![[Node.js 동기&비동기2.png]]


---
출처 - https://velog.io/@whdgh9595/TIL-%EB%8F%99%EA%B8%B0-%EB%B9%84%EB%8F%99%EA%B8%B0%EC%99%80-Node.js%EC%97%90%EC%84%9C%EC%9D%98-%EB%B9%84%EB%8F%99%EA%B8%B0%EC%B2%98%EB%A6%AC-%EB%B0%A9%EB%B2%95


https://velog.io/@octo__/%EB%8F%99%EA%B8%B0%EC%99%80-%EB%B9%84%EB%8F%99%EA%B8%B0-%EA%B7%B8%EB%A6%AC%EA%B3%A0-%EC%BD%9C%EB%B0%B1

https://it-techtree.tistory.com/entry/Nodejs-Write-Sync-Async-Code-Using-Promise-Await-Async